Change Tooling article content layout (#32841)
dotnet/AspNetCore.Docs: aspnetcore/blazor/tooling.md

---
title: Tooling for ASP.NET Core Blazor
author: guardrex
description: Learn about the tools available to build Blazor apps and how to use them.
monikerRange: '>= aspnetcore-3.1'
ms.author: riande
ms.custom: mvc
ms.date: 06/13/2024
uid: blazor/tooling
zone_pivot_groups: tooling
---
# Tooling for ASP.NET Core Blazor

[!INCLUDE[](~/includes/not-latest-version.md)]

This article describes tools for building Blazor apps using several tools:

* [Visual Studio (VS)](https://visualstudio.microsoft.com): The most comprehensive integrated development environment (IDE) for .NET developers on Windows. Includes an array of tools and features to elevate and enhance every stage of software development.

> * [Visual Studio Code (VS Code)](https://code.visualstudio.com) is an open source, cross-platform IDE that can be used to develop Blazor apps.

Reviewer: ```suggestion
* [Visual Studio Code (VS Code)](https://code.visualstudio.com) is an open source, cross-platform code editor that can be used to develop Blazor apps.
```

* [.NET CLI](/dotnet/core/tools/): The .NET command-line interface (CLI) is a cross-platform toolchain for developing, building, running, and publishing .NET applications. The .NET CLI is included with the [.NET SDK](/dotnet/core/sdk) and runs on any platform supported by the SDK.

> For assistance choosing between VS and VS Code, see the following resources:
> 
> * [What is Visual Studio?](/visualstudio/get-started/visual-studio-ide)
> * [Visual Studio Code: Getting Started](https://code.visualstudio.com/docs)
> * [Wondering which tool is best for you? We can help](https://visualstudio.microsoft.com/downloads/#picker)

Reviewer: I think the descriptions of the tooling options above is already good enough to pick the preferred option, so I think we can just remove these links.

```suggestion
```

* Select **Create**.

:::zone-end

:::zone pivot="vsc"

[Visual Studio Code](https://code.visualstudio.com) is an open source, cross-platform Integrated Development Environment (IDE) that can be used to develop Blazor apps.

Install the latest version of [Visual Studio Code](https://code.visualstudio.com/Download) for your platform.

The Visual Studio Code (VS Code) instructions for ASP.NET Core development in this tutorial use the [.NET CLI](/dotnet/core/tools/), which is part of the .NET SDK. .NET CLI commands are issued in VS Code's integrated [**Terminal**](https://code.visualstudio.com/docs/editor/integrated-terminal), which is a [PowerShell command shell](/powershell/). The **Terminal** is opened by selecting **New Terminal** from the **Terminal** menu in the menu bar.

Install the [.NET SDK](https://dotnet.microsoft.com/download). If you previously installed one or more .NET SDKs and want to see your active version, execute the following command in the **Terminal**:

> ```powershell

Reviewer: Out of curiosity, why do we prefer `powershell` to `dotnetcli` for these commands?

Author: The **Terminal** in VSC is a PowerShell prompt by default (or via a system default). They can change it to a command shell with the little drop down ...

![image](https://github.com/dotnet/AspNetCore.Docs/assets/1622880/08f8f2d9-e725-4a82-bad3-a460a6ba9390)

I'll change these to .NET CLI because that's what the commands are, and then I'll make an update to the text to say that these .NET CLI commands work at either a command prompt or a PowerShell prompt, either of which can be selected by the user with a cross-link to the VSC docs section on it.

I'll also make this change over in the Blazor tutorial because it also uses PS for these at the moment.

Author: ***Done!*** 👍 ... Will be on the next commit.

dotnet --version
```

Install the [C# Dev Kit for Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.csdevkit). For more information, see <xref:blazor/debug#visual-studio-code-prerequisites>.

If you're new to VS Code, see the [VS Code documentation](https://code.visualstudio.com/docs). If you're new to the .NET SDK, see [What is the .NET SDK?](/dotnet/core/sdk) and the associated articles in the .NET SDK documentation.

:::moniker range=">= aspnetcore-8.0"

Create a new project:

* Change to the directory using the `cd` command to where you want to create the project folder (for example, `cd c:/users/Bernie_Kopell/Documents`).

> * For a Blazor Web App experience with default interactive server-side rendering (interactive SSR), execute the following command in a command shell that uses the `blazor` project template:

Reviewer: For Visual Studio Code, there is a command pallet experience for creating project templates that we should probably use: https://code.visualstudio.com/docs/csharp/get-started#_create-a-hello-world-app

Author: This is another change that I'll apply to the Blazor tutorial PR.

Author: I don't understand part of their guidance ...

> go to the Explorer view and select Create .NET Project.

I don't see that.

I'll base our text on using the Command Palette.

Author: ***Done!*** 👍 ... Will be on the next commit.

:::moniker-end

:::zone pivot="vs"

Press <kbd>Ctrl</kbd>+<kbd>F5</kbd> on the keyboard to run the app without the debugger.

Visual Studio displays the following dialog when a project isn't configured to use SSL:

![Trust self-signed certificate dialog](~/blazor/tooling/_static/trust-certificate.png)

Select **Yes** if you trust the ASP.NET Core SSL certificate.

The following dialog is displayed:

![Security warning dialog](~/blazor/tooling/_static/install-certificate.png)

Select **Yes** to acknowledge the risk and install the certificate.

For information on trusting the development certificate for the Firefox browser, see <xref:security/enforcing-ssl#trust-the-https-certificate-with-firefox-to-prevent-sec_error_inadequate_key_usage-error>.

Visual Studio:

* Compiles and runs the app.
* Launches the default browser at `https://localhost:{PORT}`, which displays the app's UI. The `{PORT}` placeholder is the random port assigned at app creation. If you need to change the port due to a local port conflict, change the port in the project's `Properties/launchSettings.json` file.

:::zone-end

:::zone pivot="vsc"

> For information on trusting the HTTPS certificate for browsers other than Firefox, see the [HTTPS development certificate trust guidance](xref:security/enforcing-ssl#trust-the-aspnet-core-https-development-certificate-on-windows-and-macos). When using the Firefox browser, see the [certificate trust guidance for Firefox](xref:security/enforcing-ssl#trust-the-https-certificate-with-firefox-to-prevent-sec_error_inadequate_key_usage-error).

Reviewer: ```suggestion
For information on trusting the HTTPS development certificate for browsers other than Firefox, see the [HTTPS development certificate trust guidance](xref:security/enforcing-ssl#trust-the-aspnet-core-https-development-certificate-on-windows-and-macos). When using the Firefox browser, see the [certificate trust guidance for Firefox](xref:security/enforcing-ssl#trust-the-https-certificate-with-firefox-to-prevent-sec_error_inadequate_key_usage-error).
```

In VS Code, press <kbd>Ctrl</kbd>+<kbd>F5</kbd> (Windows) or <kbd>⌘</kbd>+<kbd>F5</kbd> (macOS) to run the app without debugging.

At the **Select debugger** prompt in the **Command Palette** at the top of the VS Code UI, select **C#**. At the next prompt, select the HTTPS profile (`[https]`).

The default browser is launched at `https://localhost:{PORT}`, which displays the app's UI. The `{PORT}` placeholder is the random port assigned at app creation. If you need to change the port due to a local port conflict, change the port in the project's `Properties/launchSettings.json` file.

:::zone-end

:::zone pivot="cli"

> For information on trusting the HTTPS certificate for browsers other than Firefox, see the [HTTPS development certificate trust guidance](xref:security/enforcing-ssl#trust-the-aspnet-core-https-development-certificate-on-windows-and-macos). When using the Firefox browser, see the [certificate trust guidance for Firefox](xref:security/enforcing-ssl#trust-the-https-certificate-with-firefox-to-prevent-sec_error_inadequate_key_usage-error) section of that article.

Reviewer: ```suggestion
For information on trusting the HTTPS development certificate for browsers other than Firefox, see the [HTTPS development certificate trust guidance](xref:security/enforcing-ssl#trust-the-aspnet-core-https-development-certificate-on-windows-and-macos). When using the Firefox browser, see the [certificate trust guidance for Firefox](xref:security/enforcing-ssl#trust-the-https-certificate-with-firefox-to-prevent-sec_error_inadequate_key_usage-error) section of that article.
```

  ```dotnetcli
  dotnet new blazorwasm -o BlazorApp -ho
  ```

:::moniker-end

:::moniker range="< aspnetcore-7.0"

Create a new project:

* Change to the directory using the `cd` command to where you want to create the project folder (for example, `cd c:/users/Bernie_Kopell/Documents`).

* For a Blazor WebAssembly experience, execute the following command:

  ```dotnetcli
  dotnet new blazorwasm -o BlazorApp
  ```

* For a hosted Blazor WebAssembly experience, add the hosted option (`-ho` or `--hosted`) option to the command:

  ```dotnetcli
  dotnet new blazorwasm -o BlazorApp -ho
  ```

* For a Blazor Server experience, execute the following command:

  ```dotnetcli
  dotnet new blazorserver -o BlazorApp
  ```

:::moniker-end

For more information on the templates and options, see the [Blazor project templates and template options](#blazor-project-templates-and-template-options) section.

:::zone-end

## Run the app

:::moniker range=">= aspnetcore-8.0"

> [!IMPORTANT]
> When executing a Blazor Web App, run the app from the solution's server project, which is the project with a name that doesn't end in `.Client`.

> >
> > When the app is launched, only the `Properties/launchSettings.json` file in server project is used.

Reviewer: I'm not sure why we have this mention of launchSettings.json. Is it needed?

```suggestion
```

:::moniker-end

:::moniker range="< aspnetcore-8.0"

> [!IMPORTANT]
> When executing a hosted Blazor WebAssembly app, run the app from the solution's **:::no-loc text="Server":::** project.

> >
> > When the app is launched, only the `Properties/launchSettings.json` file in the :::no-loc text="Server"::: project is used.

Reviewer: ```suggestion
```

:::moniker-end

:::zone pivot="vs"

Press <kbd>Ctrl</kbd>+<kbd>F5</kbd> on the keyboard to run the app without the debugger.

Visual Studio displays the following dialog when a project isn't configured to use SSL:

![Trust self-signed certificate dialog](~/blazor/tooling/_static/trust-certificate.png)

Select **Yes** if you trust the ASP.NET Core SSL certificate.

The following dialog is displayed:

![Security warning dialog](~/blazor/tooling/_static/install-certificate.png)

Select **Yes** to acknowledge the risk and install the certificate.

For information on trusting the development certificate for the Firefox browser, see <xref:security/enforcing-ssl#trust-the-https-certificate-with-firefox-to-prevent-sec_error_inadequate_key_usage-error>.

Visual Studio:

* Compiles and runs the app.

> * Launches the default browser at `https://localhost:{PORT}`, which displays the app's UI. The `{PORT}` placeholder is the random port assigned to the app when the app is created. If you need to change the port due to a local port conflict, you may do so in the project's `Properties/launchSettings.json` file.

Author: ```suggestion
* Launches the default browser at `https://localhost:{PORT}`, which displays the app's UI. The `{PORT}` placeholder is the random port assigned at app creation. If you need to change the port due to a local port conflict, change the port in the project's `Properties/launchSettings.json` file.
```

:::zone-end

:::zone pivot="vsc"

For information on trusting the HTTPS certificate for browsers other than Firefox, see the [HTTPS development certificate trust guidance](xref:security/enforcing-ssl#trust-the-aspnet-core-https-development-certificate-on-windows-and-macos). When using the Firefox browser, see the [certificate trust guidance for Firefox](xref:security/enforcing-ssl#trust-the-https-certificate-with-firefox-to-prevent-sec_error_inadequate_key_usage-error).

In VS Code, press <kbd>Ctrl</kbd>+<kbd>F5</kbd> (Windows) or <kbd>⌘</kbd>+<kbd>F5</kbd> (macOS) to run the app without debugging.

> At the **Select debugger** prompt in the **Command Palette** at the top of the VS Code UI, select **C#**. At the next prompt, select the HTTPS profile (`C#: BlazorWebAppMovies [https]`).

Author: ```suggestion
At the **Select debugger** prompt in the **Command Palette** at the top of the VS Code UI, select **C#**. At the next prompt, select the HTTPS profile (`[https]`).
```

> The default browser is launched at `https://localhost:{PORT}`, which displays the app's UI. The `{PORT}` placeholder is the random port assigned to the app when the app is created. If you need to change the port due to a local port conflict, you may do so in the project's `Properties/launchSettings.json` file.

Author: ```suggestion
The default browser is launched at `https://localhost:{PORT}`, which displays the app's UI. The `{PORT}` placeholder is the random port assigned at app creation. If you need to change the port due to a local port conflict, change the port in the project's `Properties/launchSettings.json` file.
```

:::zone-end

:::zone pivot="cli"

For information on trusting the HTTPS certificate for browsers other than Firefox, see the [HTTPS development certificate trust guidance](xref:security/enforcing-ssl#trust-the-aspnet-core-https-development-certificate-on-windows-and-macos). When using the Firefox browser, see the [certificate trust guidance for Firefox](xref:security/enforcing-ssl#trust-the-https-certificate-with-firefox-to-prevent-sec_error_inadequate_key_usage-error) section of that article.

In a command shell opened to the project's root folder, execute the [`dotnet watch`](/dotnet/core/tools/dotnet-watch) command to compile and start the app:

```dotnetcli
dotnet watch
```

> The app is launched at `http://localhost:{PORT}`, where the `{PORT}` placeholder is the random port assigned to the app when the app is created. If you need to change the port due to a local port conflict, you may do so in the project's `Properties/launchSettings.json` file.

Author: ```suggestion
The default browser is launched at `https://localhost:{PORT}`, which displays the app's UI. The `{PORT}` placeholder is the random port assigned at app creation. If you need to change the port due to a local port conflict, change the port in the project's `Properties/launchSettings.json` file.
```

In VS Code, press <kbd>Ctrl</kbd>+<kbd>F5</kbd> (Windows) or <kbd>⌘</kbd>+<kbd>F5</kbd> (macOS) to run the app without debugging.

At the **Select debugger** prompt in the **Command Palette** at the top of the VS Code UI, select **C#**. At the next prompt, select the HTTPS profile (`C#: BlazorWebAppMovies [https]`).

The default browser is launched at `https://localhost:{PORT}`, which displays the app's UI. The `{PORT}` placeholder is the random port assigned to the app when the app is created. If you need to change the port due to a local port conflict, you may do so in the project's `Properties/launchSettings.json` file.

:::zone-end

:::zone pivot="cli"

For information on trusting the HTTPS certificate for browsers other than Firefox, see the [HTTPS development certificate trust guidance](xref:security/enforcing-ssl#trust-the-aspnet-core-https-development-certificate-on-windows-and-macos). When using the Firefox browser, see the [certificate trust guidance for Firefox](xref:security/enforcing-ssl#trust-the-https-certificate-with-firefox-to-prevent-sec_error_inadequate_key_usage-error) section of that article.

> In a command shell opened to the project's root folder, execute the [`dotnet run`](/dotnet/core/tools/dotnet-run) command to compile and start the app:

Reviewer: Should we suggest `dotnet watch` instead so that you get hot reload support? I believe `dotnet watch` will also auto launch the browser on Windows (not sure on macOS or Linux).

Author: I suppose it was just that way for historical reasons. I'll change it here and open a follow-up issue for instances elsewhere around the docs.

Author: ***Done!*** 👍 ... Will be on the next commit.

```dotnetcli
dotnet run
```

> The default browser is launched at `http://localhost:{PORT}`, which displays the app's UI. The `{PORT}` placeholder is the random port assigned to the app when the app is created. If you need to change the port due to a local port conflict, you may do so in the project's `Properties/launchSettings.json` file.

Reviewer: Hmm, I don't think the browser will auto launch if you do `dotnet run`. I think that's only a feature of `dotnet watch`.

Author: Yes, "launched" is incorrect. It should probably say, "Open a browser at the localhost URL shown in the console." Then, it can mention that the port is randomly assigned at app generation.

Author: ***Done!*** 👍 ... Will be on the next commit.

Tooling outside of Visual Studio can interact with solution files:

* The [.NET CLI](/dotnet/core/tools/) can create solution files and list/modify the projects in solution files via the [`dotnet sln` command](/dotnet/core/tools/dotnet-sln). Other .NET CLI commands use the path of the solution file for various publishing, testing, and packaging commands.

> * [Visual Studio Code](https://code.visualstudio.com) can execute the [`dotnet sln` command](/dotnet/core/tools/dotnet-sln) and other [.NET CLI commands](/dotnet/core/tools/) through its integrated **Terminal** that interact with the solution file's contents, but VS Code doesn't use the settings in a solution file directly.

Reviewer: Hmm, is this correct? I just tried out adding a project to an existing .NET solution using the Solution Explorer in C# Dev Kit and it updated the solution file accordingly.

Author: Yes, it's stale content now. I'll remove this remark.

```suggestion
```

For more information on Blazor project templates, see <xref:blazor/project-structure>.

:::moniker range=">= aspnetcore-8.0"

Rendering terms and concepts used in the following subsections are introduced in the following sections of the *Fundamentals* overview article:

* [Client and server rendering concepts](xref:blazor/fundamentals/index#client-and-server-rendering-concepts)
* [Static and interactive rendering concepts](xref:blazor/fundamentals/index#static-and-interactive-rendering-concepts)
* [Render modes](xref:blazor/fundamentals/index#render-modes)

Detailed guidance on render modes is provided by the <xref:blazor/components/render-modes> article.

### Interactive render mode

> <!-- This section and the next have been generalized by going
>      with the words "use" and "using" over "select" and "selecting" 
>      because this will now apply to all tooling types. -->
> 

Author: ```suggestion
```

* [Introduction to projects and solutions](/visualstudio/get-started/tutorial-projects-solutions)
* [What are solutions and projects in Visual Studio?](/visualstudio/ide/solutions-and-projects-in-visual-studio)

> ## Blazor template options

Reviewer: Above we go into a lot of detail on the different Blazor template options. Instead of showing all the variants, should we instead just show the default with no options and then reference the content in this section for details on all the various options?

Author: Some readers will be slightly annoyed by being thrown to a different section, but I'll change it.

Also, the template options section will need to receive the options content because it only lists the templates and recommends the *Project Structure* article or `-h|--help` for options details. I'll move at least some of the content down to this section, and I'll try to condense it a bit. Perhaps, a table would be a good choice for the content. I'll see what I can work out.

Author: ***Done!*** 👍 ... Will be on the next commit.
